aws/resource_aws_iot_certificate.go

```diff
@@ -28,26 +28,58 @@ func resourceAwsIotCertificate() *schema.Resource {
 				Type:     schema.TypeString,
 				Computed: true,
 			},
+			"certificate_pem": {
+				Type:      schema.TypeString,
+				Computed:  true,
+				Sensitive: true,
+			},
+			"public_key": {
+				Type:      schema.TypeString,
+				Computed:  true,
+				Sensitive: true,
+			},
+			"private_key": {
+				Type:      schema.TypeString,
+				Computed:  true,
+				Sensitive: true,
+			},
 		},
 	}
 }
 
 func resourceAwsIotCertificateCreate(d *schema.ResourceData, meta interface{}) error {
 	conn := meta.(*AWSClient).iotconn
 
-	log.Printf("[DEBUG] Creating certificate from csr")
-	out, err := conn.CreateCertificateFromCsr(&iot.CreateCertificateFromCsrInput{
-		CertificateSigningRequest: aws.String(d.Get("csr").(string)),
-		SetAsActive:               aws.Bool(d.Get("active").(bool)),
-	})
+	if d.Get("csr") != "" {
+		log.Printf("[DEBUG] Creating certificate from csr")
+		out, err := conn.CreateCertificateFromCsr(&iot.CreateCertificateFromCsrInput{
+			CertificateSigningRequest: aws.String(d.Get("csr").(string)),
+			SetAsActive:               aws.Bool(d.Get("active").(bool)),
+		})
 
-	if err != nil {
-		log.Printf("[ERROR] %s", err)
-		return err
-	}
-	log.Printf("[DEBUG] Created certificate from csr")
+		if err != nil {
+			log.Printf("[ERROR] %s", err)
+			return err
+		}
+		log.Printf("[DEBUG] Created certificate from csr")
 
-	d.SetId(*out.CertificateId)
+		d.SetId(*out.CertificateId)
+	} else {
+		log.Printf("[DEBUG] Creating keys and certificate")
+		out, err := conn.CreateKeysAndCertificate(&iot.CreateKeysAndCertificateInput{
+			SetAsActive: aws.Bool(d.Get("active").(bool)),
+		})
+
+		if err != nil {
+			log.Printf("[ERROR] %s", err)
+			return err
```

Comment:
Same here regarding the log message and error context 👍 

```suggestion
			return fmt.Errorf("error creating IoT Keys and Certificate: %s", err)
```